File: Editor/EasyRPG.Editor.Serializer/RM2K/LDB/LdbBattleCommands.cs
using EasyRPG.Editor.RPG;

namespace EasyRPG.Editor.Serializer.RM2K.LDB
{
	public partial class LdbReader
	{
		#region BehaviorMethods
		private BattleCommands ReadBattleCommands()
		{
			var battleCommands = new BattleCommands();

			while (!stream.EndOfFile())
			{
				var chunkId = (ChunkBattleCommands)stream.ReadCompressedInt();
				if (chunkId == (int)ChunkData.End)
				{
					break;
				}

				var chunkLenght = stream.ReadCompressedInt();
				if (chunkLenght == 0)
				{
					continue;
				}

				switch (chunkId)
				{
					case ChunkBattleCommands.Command:
						for (var i = stream.ReadCompressedInt(); i > 0; i--)
						{
							battleCommands.Commands.Add(ReadBattleCommand());
						}
						break;
					case ChunkBattleCommands.Placement:
						battleCommands.Placement = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.Row:
						battleCommands.Row = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.BattleType:
						battleCommands.BattleType = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.DeathHandler1:
						battleCommands.DeathHandler1 = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.Unknown1:
						battleCommands.Unknown1 = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.DeathHandler2:
						battleCommands.DeathHandler2 = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.DeathEvent:
						battleCommands.DeathEvent = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.WindowSize:
						battleCommands.WindowSize = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.Transparency:
						battleCommands.Transparency = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.Teleport:
						battleCommands.Teleport = stream.ReadCompressedBool();
						break;
					case ChunkBattleCommands.TeleportId:
						battleCommands.TeleportId = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.TeleportX:
						battleCommands.TeleportX = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.TeleportY:
						battleCommands.TeleportY = stream.ReadCompressedInt();
						break;
					case ChunkBattleCommands.TeleportFace:
						battleCommands.TeleportFace = stream.ReadCompressedInt();
						break;
					default:
						stream.Skip(chunkLenght);
						break;
				}
			}
			return battleCommands;
		}
		#endregion
	}
}

File: readers/include/detail/ldb_battlecommands.cpp
/////////////////////////////////////////////////////////////////////////////
// This file is part of EasyRPG.
//
// EasyRPG is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// EasyRPG is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
/////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////
// Headers
////////////////////////////////////////////////////////////
#include "ldb_reader.h"
#include "ldb_chunks.h"
#include "reader_lcf.h"
#include "reader_struct.h"

////////////////////////////////////////////////////////////
/// Read BattleCommands
////////////////////////////////////////////////////////////
EASYRPG_STRUCT_ID_READER(BattleCommands, NoID)

EASYRPG_STRUCT_NAME(BattleCommands)

#define EASYRPG_CHUNK_SUFFIX LDB_Reader
#define EASYRPG_CURRENT_STRUCT BattleCommands

EASYRPG_STRUCT_FIELD_MAP_BEGIN(BattleCommands)
	, EASYRPG_TYPED_FIELD_PAIR(int, placement, ('plac')('emen')('t'))
	, EASYRPG_TYPED_FIELD_PAIR(int, death_handler1, ('deat')('h_ha')('ndle')('r1'))
	, EASYRPG_TYPED_FIELD_PAIR(int, row, ('row'))
	, EASYRPG_TYPED_FIELD_PAIR(int, battle_type, ('batt')('le_t')('ype'))
	, EASYRPG_TYPED_FIELD_PAIR(int, unknown1, ('unkn')('own1'))
	, EASYRPG_TYPED_FIELD_PAIR(std::vector<RPG::BattleCommand>, commands, ('comm')('ands'))
	, EASYRPG_TYPED_FIELD_PAIR(int, death_handler2, ('deat')('h_ha')('ndle')('r2'))
	, EASYRPG_TYPED_FIELD_PAIR(int, death_event, ('deat')('h_ev')('ent'))
	, EASYRPG_TYPED_FIELD_PAIR(int, window_size, ('wind')('ow_s')('ize'))
	, EASYRPG_TYPED_FIELD_PAIR(int, transparency, ('tran')('spar')('ency'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, teleport, ('tele')('port'))
	, EASYRPG_TYPED_FIELD_PAIR(int, teleport_id, ('tele')('port')('_id'))
	, EASYRPG_TYPED_FIELD_PAIR(int, teleport_x, ('tele')('port')('_x'))
	, EASYRPG_TYPED_FIELD_PAIR(int, teleport_y, ('tele')('port')('_y'))
	, EASYRPG_TYPED_FIELD_PAIR(int, teleport_face, ('tele')('port')('_fac')('e'))
EASYRPG_STRUCT_FIELD_MAP_END()

#undef EASYRPG_CURRENT_STRUCT
#undef EASYRPG_CHUNK_SUFFIX
